### PR TITLE
Avoid with overhead in memory allocation

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

**Comment by okuta** [comment]
Jenkins, test this please

**Comment by takagi** [comment]
Would you resolve the conflict?

**Comment by niboshi** [comment]
`cupy.cuda.memory.alloc(10)`
Before the fix: 0.96 us
After the fix: 0.75 us

**Comment by takagi** [comment]
It looks fine. I also checked the CPP file that Cython generates and found that the number of touching reference counters is reduced.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by okuta** [comment]
I resolved conflicts.

**Comment by niboshi** [comment]
@takagi Let me review the change.

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -845,6 +845,30 @@ cpdef size_t _bin_index_from_size(size_t size):
...
 
+cdef _gc_isenabled = gc.isenabled
+cdef _gc_disable = gc.disable
+cdef _gc_enable = gc.enable
+
+
+cdef bint _lock_no_gc(lock):
+    rlock.lock_fastrlock(lock, -1, True)
+
+    # This method may be called from the context of finalizer
```
Comment (niboshi) [inline cupy/cuda/memory.pyx]:
```suggestion
    # This function may be called from the context of finalizer
```

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -845,6 +845,30 @@ cpdef size_t _bin_index_from_size(size_t size):
     return (size - 1) // ALLOCATION_UNIT_SIZE
 
 
+cdef _gc_isenabled = gc.isenabled
+cdef _gc_disable = gc.disable
+cdef _gc_enable = gc.enable
+
+
+cdef bint _lock_no_gc(lock):
```
Comment (niboshi) [inline cupy/cuda/memory.pyx]: Please briefly document the return value.

Comment (okuta) [reply cupy/cuda/memory.pyx]: I added.

**Comment by niboshi** [comment]
Thanks!
Jenkins, test this please

**Pull request merged**